cfspeedtest/cloudflare.py

```diff
@@ -0,0 +1,249 @@
+"""
+Library for the Cloudflare speedtest suite.
+
+This uses endpoints from speed.cloudflare.com.
+"""
+
+from __future__ import annotations
+
+import statistics
+import time
+from enum import Enum
+from typing import TYPE_CHECKING, Any, NamedTuple
+
+import requests
+
+if TYPE_CHECKING:
+    from logging import Logger
+
+
+class TestType(Enum):
+    """The type of an individual test."""
+
+    Down = "GET"
+    Up = "POST"
+
+
+class TestSpec(NamedTuple):
+    """The specifications of an individual test."""
+
+    size: int
+    """The size of the test in bytes."""
+    iterations: int
+    name: str
+    type: TestType
+
+    @property
+    def bits(self) -> int:
+        """The size of the test in bits."""
+        return self.size * 8
+
+
+TestSpecs = tuple[TestSpec, ...]
+
+DOWNLOAD_TESTS: TestSpecs = (
+    TestSpec(100_000, 10, "100kB", TestType.Down),
+    TestSpec(1_000_000, 8, "1MB", TestType.Down),
+    TestSpec(10_000_000, 6, "10MB", TestType.Down),
+    TestSpec(25_000_000, 4, "25MB", TestType.Down),
+)
+UPLOAD_TESTS: TestSpecs = (
+    TestSpec(100_000, 8, "100kB", TestType.Up),
+    TestSpec(1_000_000, 6, "1MB", TestType.Up),
+    TestSpec(10_000_000, 4, "10MB", TestType.Up),
+)
+DEFAULT_TESTS: TestSpecs = (
+    TestSpec(1, 20, "latency", TestType.Down),
+    *DOWNLOAD_TESTS,
+    *UPLOAD_TESTS,
+)
+
+
+class TestResult(NamedTuple):
+    """The result of an individual test."""
+
+    value: Any
+    time: float = time.time()
+
+
+class TestTimers(NamedTuple):
+    """A collection of test timer collections, measured in seconds."""
+
+    full: list[float]
+    """The times taken to prepare and perform the requests."""
+    server: list[float]
+    """The times taken to process the requests as reported by the worker."""
+    request: list[float]
+    """The internal client times elapsed to complete the requests."""
+
+    def to_speeds(self, test: TestSpec) -> list[int]:
+        """Compute the test speeds in bits per second from its type and size."""
+        if test.type == TestType.Up:
+            return [int(test.bits / server_time) for server_time in self.server]
+        return [
+            int(test.bits / (full_time - server_time))
+            for full_time, server_time in zip(self.full, self.server)
+        ]
+
+    def to_latencies(self) -> list[float]:
+        """Compute the test latencies in milliseconds."""
+        return [
+            (request_time - server_time) * 1e3
+            for request_time, server_time in zip(self.request, self.server)
+        ]
+
+    @staticmethod
+    def jitter_from(latencies: list[float]) -> float | None:
+        """Compute jitter as average deviation between consecutive latencies."""
+        if len(latencies) < 2:
+            return None
+        return statistics.median(
```

Comment:
while i was moving these, it occurred to me that using medians is odd. typically, arithmetic means are used for network statistics, especially when [measuring jitter](https://github.com/cloudflare/speedtest/blob/307617edcf57d2bffa6fe4b89bd223a1c26f0ba5/src/Results/MeasurementCalculations.js#L17). should i change these to use the mean instead of the median?

Reply:
I think you're right. The intention of measuring jitter is to capture the average variability in latency.
If we use the mean, we lose accuracy.

I double-checked the Cloudflare approach, and they also use the arithmetic mean.
In effect, it would be sensible to change this to `statistics.mean`.